File: src/main/java/com/github/simiacryptus/aicoder/openai/translate/BaseTranslationRequest.java
package com.github.simiacryptus.aicoder.openai.translate;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public abstract class BaseTranslationRequest<T extends BaseTranslationRequest<T>> implements TranslationRequest {

    private CharSequence inputTag;
    private CharSequence outputTag;
    private CharSequence instruction;
    @NotNull
    private final Map<CharSequence, CharSequence> inputAttr = new HashMap<>();
    @NotNull
    private final Map<CharSequence, CharSequence> outputAttr = new HashMap<>();
    private CharSequence originalText;
    private double temperature;
    private int maxTokens;

    @Override
    public @NotNull String getInputTag() {
        return inputTag.toString();
    }

    @Override
    public @NotNull String getOutputTag() {
        return outputTag.toString();
    }

    @Override
    public @NotNull CharSequence getInstruction() {
        return instruction.toString();
    }

    @Override
    @NotNull
    public Map<CharSequence, CharSequence> getInputAttr() {
        return Collections.unmodifiableMap(inputAttr);
    }

    @Override
    @NotNull
    public Map<CharSequence, CharSequence> getOutputAttr() {
        return Collections.unmodifiableMap(outputAttr);
    }

    @Override
    public @NotNull String getOriginalText() {
        return originalText.toString();
    }

    @Override
    public double getTemperature() {
        return temperature;
    }

    @Override
    public int getMaxTokens() {
        return maxTokens;
    }

    @Override
    public @NotNull T setInputType(CharSequence inputTag) {
        this.inputTag = inputTag;
        return (T) this;
    }

    @Override
    public @NotNull T setOutputType(CharSequence outputTag) {
        this.outputTag = outputTag;
        return (T) this;
    }

    @Override
    public @NotNull T setInstruction(CharSequence instruction) {
        this.instruction = instruction;
        return (T) this;
    }

    @Override
    public @NotNull T setInputText(CharSequence originalText) {
        this.originalText = originalText;
        return (T) this;
    }

    @Override
    public @NotNull T setTemperature(double temperature) {
        this.temperature = temperature;
        return (T) this;
    }

    @Override
    public @NotNull T setMaxTokens(int maxTokens) {
        this.maxTokens = maxTokens;
        return (T) this;
    }

    @Override
    public @NotNull T setInputAttribute(CharSequence key, @Nullable CharSequence value) {
        if(null == value || value.length()==0) {
            inputAttr.remove(key);
        } else {
            inputAttr.put(key, value);
        }
        return (T) this;
    }

    @Override
    public @NotNull T setOutputAttrute(CharSequence key, @Nullable CharSequence value) {
        if(null == value || value.length()==0) {
            outputAttr.remove(key);
        } else {
            outputAttr.put(key, value);
        }
        return (T) this;
    }

}

File: src/main/kotlin/com/github/simiacryptus/aicoder/openai/translate/BaseTranslationRequest.kt
package com.github.simiacryptus.aicoder.openai.translate

import com.github.simiacryptus.aicoder.config.AppSettingsState

abstract class BaseTranslationRequest<T : BaseTranslationRequest<T>>(settings: AppSettingsState) : TranslationRequest {
    init {
        setTemperature(settings.temperature)
        setMaxTokens(settings.maxTokens)
    }

    override var inputTag: CharSequence? = null
    override var outputTag: CharSequence? = null
    override var instruction: CharSequence? = null
    override val inputAttr: MutableMap<CharSequence?, CharSequence> = HashMap()
    override val outputAttr: MutableMap<CharSequence?, CharSequence> = HashMap()

    override var originalText: CharSequence? = null
    override var temperature = 0.0
    override var maxTokens = 0

    override fun setInputType(inputTag: CharSequence?): T {
        this.inputTag = inputTag
        return this as T
    }

    override fun setOutputType(outputTag: CharSequence?): T {
        this.outputTag = outputTag
        return this as T
    }

    override fun setInstruction(instruction: CharSequence?): T {
        this.instruction = instruction
        return this as T
    }

    override fun setInputText(originalText: CharSequence?): T {
        this.originalText = originalText
        return this as T
    }

    override fun setTemperature(temperature: Double): T {
        this.temperature = temperature
        return this as T
    }

    override fun setMaxTokens(maxTokens: Int): T {
        this.maxTokens = maxTokens
        return this as T
    }

    override fun setInputAttribute(key: CharSequence?, value: CharSequence?): T {
        if (null == value || value.length == 0) {
            inputAttr.remove(key)
        } else {
            inputAttr[key] = value
        }
        return this as T
    }

    override fun setOutputAttrute(key: CharSequence?, value: CharSequence?): T {
        if (null == value || value.length == 0) {
            outputAttr.remove(key)
        } else {
            outputAttr[key] = value
        }
        return this as T
    }
}